### DISCUSSION TITLE
mdit container 选择器语法不支持嵌套，以及分栏语法的改进建议

### DISCUSSION THREAD

**Discussion by YiwangHuang** [any-block/any-block#128] [General]
## BUG 提交

mdit container 选择器语法不支持嵌套，因为看到作者个人网站上的说明[Selector](https://linczero.github.io/MdNote_Public/ProductDoc/AnyBlock/docs/en/03.%20Selector.html#summary)才去尝试的，不知道是还没来得及实现，还是某次更新后出的问题。

## 改进建议

1. 对于 mdit container 选择器的下 col 分栏格式，能否支持将`@col`后的同一行的内容作为参数控制各分栏占比，我认为非等分可控比例的分栏对插图很有帮助。我希望如下的 md 文本能被渲染成 1:1:2 的三分栏
   
```md
:::col

@col 25

colText 1 

@col 25

colText 2

@col 50

colText 3

:::
```

2. 优化分栏的渲染风格，目前分栏间距太大了，我认为是`.ab-col`类的`gap`属性和`.ab-items-item`类中的`padding`属性在功能上重复了。

> PS: 这两项改进设计的代码比较简单，我已经通过二次开发基本实现了，如果作者觉得合适的话我可以再完善一下，做更多测试后提交

我实现根据参数控制分栏占比，并把 `.ab-col` 类的 `gap` 属性删掉后的渲染结果
![image](https://github.com/user-attachments/assets/e1a42aef-27f8-4ca0-9a7a-c911d5ce6816)

**Comment by LincZero** [comment]
## markdown-it-container语法嵌套问题（BUG）

我知道这个问题，但没来得及实现：

头尾选择器（markdown-it-container语法和代码围栏块语法）当前存在的嵌套问题：

1. 代码围栏块支持 \`\`\` 和 `~~~` 的二层嵌套，未支持 \`\`\`\` 嵌套 \`\`\` 这种多层嵌套语法
2. markdown-it-contianer 未支持 `::::` 嵌套 `:::` 这种多层嵌套语法
3. 同类嵌套问题。
    当使用 `::::` 嵌套 `:::` 时，无法是相同类型的markdown-it-container类型，例如不能tab嵌套tab、col不能嵌套col，只能tab嵌套col这种类型不同的（因为@col和@tab标识不能区分到底是哪一层容器的标注）。
    但这一问题似乎是语法设计的问题，vuepress/vitepress本身就存在这一问题，所以我并不爱用这个东西的嵌套，他本身就有语法问题没有解决（我感觉 `::::` 应该用 `@@` 这样标注区分开会更合理，但他原生并没这种设计）。一般我是markdown-it-container去嵌套标题/列表/代码块选择器这样会比较多
    
这个东西你可以PR，代码块围栏那里其实逻辑一样，方便的话也可以顺便修一下

## 分栏问题（扩展）

先进行一点设计讨论吧：

这种设计是原生语法吗？宽度信息体现在标题上会不会不太好？

我之前的一个设计解是自定义样式，例如 `[addClass(ab-custom-td1-100)]`、`[addClass(ab-custom-td1-half)]` 可以控制第一列的宽度（把`1`删掉就是控制每列的宽度，已实装），可以应用到任意表格（主要是针对list2table、dir、list2lt等表，你可以试一下体验下）。
除了自带的，用户也可以自己自定义样式，然后使用 `addClass` 处理器来匹配预设样式。

不过用class的不够灵活，预期的后续改进是可以新增一个处理器，用来控制宽度，例如： `:::col|width(25-25-50)` 这样，width处理器用于添加带nth选择器的style（实现参考addClass、scroll、fold等处理器），这样对于表格也能起效

**Comment by YiwangHuang** [comment]
## markdown-it-container语法嵌套问题

这部分我只是目前只提一个问题。改成支持嵌套是不是涉及渲染逻辑，等放假了可以好好看看你网站上的开发笔记尝试解决。

## 分栏问题（扩展）

目前我主要使用了 markdown-it-container 的 col、tab 和 card 功能。在搭建网站时，我使用的是 Quarto，它的双栏语法类似于 Pandoc 的实现方式：
```md
:::: {.columns}
::: {.column width="40%"}
Left column
:::
::: {.column width="60%"}
Right column
:::
::::
```
这种分栏功能实际上是通过嵌套来实现的。不过，我也在考虑将网站迁移到 VuePress。

但是我觉得你说的有道理。我目前想到两种可行方案，它们都能在语法上保持向前兼容：

1. 你说的加一个选择器，类似`:::col|width(25-25-50)`
2. 在`@block`后面加参数，类似`@block text | width(25%)`，感觉更灵活，但是就只能支持markdwon-it-container下col和card。

你更倾向于哪种？关于语法扩展这块，我想近期pr一下

**Comment by LincZero** [comment]
## markdown-it-container语法嵌套问题

你代码是改了分栏，没改mdit-container嵌套是吗？这部分的修改可以参考已实现的代码块 \`\`\` 和 `~~~` 的嵌套，逻辑是一样的。可参考：`src/ab_manager/abm_cm/ABSelector_MdBase.ts::173行`

## 分栏问题（扩展）

我倾向于添加一个新的处理器。原因：

1. 这样设计有更多的应用点：一是表格/卡片都可应用、二是除mdit-container外的其他anyblock选择器也可应用，而非只支持markdown-it-container
2. 我认为表格的可控制宽度有一定的需求度。例如配合表格选择器，感觉还不错
   ```md
   [width(25,75)]
   
   |1|2|
   |---|---|
   |3|4|
   ```
3. 只是增加处理器罢了，完全没有引入新的规则，语法设计和架构完全没变，完全向前兼容
4. 程序修改相对较简单，可参考：`src/ABConverter/converter/abc_deco.ts::75行、109行`。新的处理器应该放这，html转html的装饰处理器应该放这个文件里

而你说的Quarto的这个语法，我似乎没有在vuepress中看到（如果存在较为规范和通用的类似语法，再进行一遍考虑），似乎不是mdit-container中较通用的语法

---

语法设计补充：
- `width(25%,25%,50%)` 、`witdh(100px, 200px)` // 允许有3列但仅手动指定前几列
- 单位可以省略，省略单位感觉是px合理些？

**Comment by YiwangHuang** [comment]
OK，感谢指路，我先尝试按你的思路新增一个选择器，省略单位就按px好了。

**Comment by YiwangHuang** [comment]
我新增了宽度控制处理器，提交了PR。后续寒假我再改markdown-it-container语法嵌套问题，settingtab也可以帮忙改成分页的形式。

问个问题，这个项目可以直接适用于 vuepress 吗？没用过vuepross不太懂

**Comment by LincZero** [comment]
看到了，在检查了，暂时看起来没啥问题，我自己安装用一下先。

我写成markdown-it插件的形式，教程/demo可以看any-block组织下的另外两个项目（vitepress-demo 和 vuepress-demo），其他支持mdit的博客框架理论也可以但我没试过

settingtab 目前感觉比较鸡肋，因为这个插件感觉没什么好设置的。目前唯一的设置项是自定义别名，但感觉用的人不多就没优化。其他显示内容大多作为一个 “文档补充” 来使用，本质上也是文档作用

**Comment by YiwangHuang** [comment]
补充私有格式这块，比插件市场的很多高下载量插件都方便。不过，文档有点极客，学习成本有点高。而且很多 Ob 用户其实根本不会点开外链看，我主要想在设置界面中完善一下功能介绍（操作GIF也加一下），方便推给同事朋友用。

**Comment by LincZero** [comment]
目前想到的降低入门门槛的两个设计：

## 功能介绍、使用与教程一体问题

包括你说的在设置面板上进行部分介绍，而不是在链接中。

但是篇幅太大了，塞不下设置面板吧。不想点外链的话，设置面板支持内部iframe加载链接，或用按钮在ob内打开说明界面，可能还好些

我用过比较理想的使用和教程一体的软件：（可供进行设计参考）

一是QtCreator等一些内部IDE，QtCreator自带API文档，可IDE中遇到某个接口可以进原型，也可以跳转到对应的API文档页。
二是Houdini，自带每种节点的文档，某个节点不会时，可以直接跳转到节点对应的文档页。

## 引导式创建、使用

引导式创建可以是有个菜单可以快速创建，例如 admonition 的一个命令可以帮助创建 callout 块。例如 anyblock 增加右键菜单或命令式创建 ab块，最好是那种有专门界面和效果展示的菜单，类似notion块那种效果

## 文档改善

至于你说的 “文档有点极客”，我感觉进入效果展示页后，我感觉其实入门还好吧……
高级和很多细节的东西我确实没写太详细、或者干脆没写。文档不太好的地方往往是入门往后的，没有时间完善的地方

而且东西太多了，感觉现在比较像mermaid和plantuml官网文档那种情况

**Comment by YiwangHuang** [comment]
等假期里都可以尝试一下。我认为介绍界面最需要增加的是效果展示GIF，主要这个添加起来也最方便。ob插件里，latex suite 的 GIF 展示做得特别好，虽然使用门槛也不低，但一看GIF就很有尝试欲。设置面板内容多了，分页也有必要了。

**Comment by YiwangHuang** [comment]
@LincZero
关于嵌套问题，我发现仅修改`src/ab_manager/abm_cm/ABSelector_MdBase.ts`是不够的，只能让头尾选择器正确匹配同层级的`:::`并返回之间的内容。但是插件不会识别 `@block` 下有没有mdit块（但是部分其他选择器语法可以被识别），所以内层嵌套的mdit块会被当成普通的md文本处理。但我不太懂渲染逻辑，没找到相关代码位置，方便指点一下吗？

**Comment by LincZero** [comment]
哦对，我想漏了。

1. 在实时模式的**初次**ab渲染 (第一层) 走的是CM中的选择器
2. 在阅读模式中 *初次*ab渲染  (第一层) 走的是HTML中的选择器的一个分支(b2)
3. 被嵌套内容的再次渲染，前者一定是HTML内容，所以走的都是HTML中的选择器的另一个分支(b1)

（ob插件的操蛋之处，碰到这种东西要两倍的开发量。实时模式一遍，阅读模式一遍）

原因是被嵌套的内容的进行 “重新渲染” 时，会直接渲染成HTML内容。即需要让 `:::` 在阅读模式及“重新渲染”中被识别。
对应的代码是 `src/ab_manager_abm_html/ABSelector_PostHTML.ts`。逻辑大概是：从42行的b1分支 -> 185行的findABBlock_recurve

然后这个函数只识别了anyblock的 `[]` 形式的选择器，其实是我之前是没有做 `:::` 的重渲染的嵌套识别……
**这个其实有点不好做**，需要在html结构中去识别他是否是一个 `:::` 结构

**Comment by LincZero** [comment]
其实很难搞，按我上面说的话。想到的另一个方法是直接用markdown-it引擎来解析（markdown-it自带文档解析环节，比我在ob实时模式和阅读模式手搓的手动解析完善和舒服太多了，可以轻松扩展markdown-it插件）。这个方法优点：**实现更简单、更稳定、可以进一步扩展markdown-it的插件、可以单开插件解耦**。如果你会markdown-it的开发的话，我推荐用这个方法。并且如果使用这个方法，也可以另开一个插件。

思路：import个markdown-it包，碰到 `:::` 直接走markdown-it的解析规则。他解析结束时会将这个 `:::` 变成一个特殊的 token 项，然后可以将这个token项内的内容再替换成ob的渲染内容（以便内容项使用ob的渲染方式，避免ob的一些行为和插件不生效）

有些插件也这样做，像老旧的 table-extend，本身就是复用了 markdown-it-table-extend！

**Comment by LincZero** [comment]
第二种方式的 “单独开一个插件” 详细说一下：

开一个新插件，效果将代码块内容使用markdown-it引擎进行渲染（部分局部内容用ob渲染）（只需要实现在代码块内解析和渲染就行了，不用关实时和阅读模式下代码块外的识别）

````
```markdown-it
:::: ...
::: ……
……
::: ……
……
::::

|table-extend等mdit插件|...|
|---|---|
|    |    |
```
````

实现方式：用markdown-it包进行解析、处理。正文部分用回ob的渲染函数

然后通过与anyblock联动打通不使用代码块方式的渲染，类比这种用法：https://linczero.github.io/MdNote_Public/ProductDoc/AnyBlock/docs/zh/05.%20%E7%B2%BE%E5%BD%A9%E7%94%A8%E4%BE%8B.html#mindmap-nexgen

我觉得这种方式其通用、解耦、易开发性、稳定性更好

**Comment by YiwangHuang** [comment]
谢谢提供的思路！不过我这方面不太熟，估计最近搞不定，等过两个月我再来尝试。这段时间我先用 VuePress 把自己的网站重建一下，顺便熟悉熟悉 Markdown-it，到时候应该会更好上手。

**Comment by LincZero** [comment]
好的。网站的**粗略**搭建可以参考：https://github.com/LincDocs/Workflow ，实例参考：https://github.com/PKM-er/Pkmer-Math

只需要两步：添加action文件、开启page选项。效果同我网站，工作流支持anyblock语法、支持双链

如果想要更精细的控制和自定义，还得单独创建vuepress/vitepress项目 (参考any-block组织下的另外两个项目)。快速搭建方案只是零配置，方便可以先临时用着

---

Markdwon-it以及obsidian等的学习，我这里有一些我学习时的笔记，希望可以提供参考：https://linczero.github.io/MdNote_Public/01.%20DesignAndDevelop/Develop/04.%20Project/Type/Doc/

**Comment by LincZero** [comment]
close，使用 #160 代替
